lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -0,0 +1,154 @@
+module: package_facts
+short_description: package information as facts
+description:
+    - Return information about installed packages as facts
+options:
+    manager:
+      description:
+        - The package manager used by the system so we can query the package information
+      default: autodetect
+      required: False
+version_added: "2.5"
+author:
+    - Matthew Jones
+    - Brian Coca
+'''
+
+EXAMPLES = '''
+    - name: get the package facts
+      package_facts:
+
+    - name: show them
+      debug: msg="{{ansible_facts.packages}}"
+'''
+
+RETURN = '''
+ansible_facts:
+  type: dict
+  contains:
+      packages:
+      type: dict
+      contains: list of dicts with package information
+      sample: [
+                {
+                 "libbz2-1.0": [
+                    {
+                      "version": "1.0.6-5",
+                      "source": "apt",
+                      "arch": "amd64",
+                      "name": "libbz2-1.0"
+                    }
+                  ],
+                 "patch": [
+                    {
+                      "version": "2.7.1-4ubuntu1",
+                      "source": "apt",
+                      "arch": "amd64",
+                      "name": "patch"
+                    }
+                  ],
+                }
+              ]
+'''
+
+import sys
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+
+
+def rpm_package_list():
+
+    try:
+        import rpm
+    except ImportError:
+        module.fail_json('Unable to use the rpm python bindings, please ensure they are installed under the python the module runs under')
+
+    trans_set = rpm.TransactionSet()
+    installed_packages = {}
+    for package in trans_set.dbMatch():
+        package_details = dict(name=package[rpm.RPMTAG_NAME],
+                               version=package[rpm.RPMTAG_VERSION],
+                               release=package[rpm.RPMTAG_RELEASE],
+                               epoch=package[rpm.RPMTAG_EPOCH],
+                               arch=package[rpm.RPMTAG_ARCH],
+                               source='rpm')
+        if package_details['name'] not in installed_packages:
+            installed_packages[package_details['name']] = [package_details]
+        else:
+            installed_packages[package_details['name']].append(package_details)
+    return installed_packages
+
+
+def deb_package_list():
+
+    try:
+        import apt
+    except ImportError:
+        module.fail_json('Unable to use the apt python bindings, please ensure they are installed under the python the module runs under')
+
+    apt_cache = apt.Cache()
+    installed_packages = {}
+    apt_installed_packages = [pk for pk in apt_cache.keys() if apt_cache[pk].is_installed]
+    for package in apt_installed_packages:
+        ac_pkg = apt_cache[package].installed
+        package_details = dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture, source='apt')
```

Comment:
It would be easier if RPM and DEB were to give the same structured reply, or having an agnostic module would not be very useful (in cases like "if upstream version is greater than x.y"). Comparing versions is a complex operation and there is no filter around `dpkg --compare-versions`.

Also having the full version (epoch+upstream+release, following the standard distro notation) along with the upstream version would be useful I think.